File: frontend/src/types/valorant.ts
export interface ValorantMatchState {
  matchId: string;
  mapName: string; // e.g., "Ascent", "Haven", "Lotus"
  mode: string; // e.g., "Competitive", "Premier"
  serverRegion: string; // e.g., "AP - Mumbai", "NA - N. Virginia"
  playerTeam: "RED" | "BLUE";
  teamScore: number;
  enemyScore: number;
  roundNumber: number;
  phase: "BUY_PHASE" | "COMBAT" | "POST_ROUND";
}

export interface PlayerRealtimeStats {
  puuid: string;
  riotId: string; // e.g., "Player#AP1"
  agentName: string; // e.g., "Jett", "Omen", "Viper"
  agentIconUrl: string;
  kills: number;
  deaths: number;
  assists: number;
  kdRatio: number;
  combatScore: number;
  economyCredits: number;
  currentTierName: string; // e.g., "Immortal 1"
  currentTierIconUrl: string;
  rankingRating: number; // 0 - 100 RR inside current rank
  rrChangeLastMatch: number; // e.g., +21 or -14
  mapWinRate: number; // Percentage win rate on the currently loaded map
  historicalMatchesOnMap: number;
}

export interface OverlayTelemetryPayload {
  timestamp: number;
  matchState: ValorantMatchState;
  playerStats: PlayerRealtimeStats;
}

File: backend/internal/model/valorant.go
package model

import "time"

// --- Raw Riot API Representation (Input Data) ---

type RiotAccountResponse struct {
	PUUID    string `json:"puuid"`
	GameName string `json:"gameName"`
	TagLine  string `json:"tagLine"`
}

type ActiveShardResponse struct {
	PUUID string `json:"puuid"`
	Game  string `json:"game"`
	ActiveShard string `json:"activeShard"`
}

type RawMatchListResponse struct {
	PUUID   string   `json:"puuid"`
	Matches []string `json:"history"`
}

// --- Pruned Live HUD Telemetry Protocol (Sub-Kilobyte Output Data) ---

type ValorantMatchState struct {
	MatchID      string `json:"matchId"`
	MapName      string `json:"mapName"`
	Mode         string `json:"mode"`
	ServerRegion string `json:"serverRegion"`
	PlayerTeam   string `json:"playerTeam"` // "RED" | "BLUE"
	TeamScore    int    `json:"teamScore"`
	EnemyScore   int    `json:"enemyScore"`
	RoundNumber  int    `json:"roundNumber"`
	Phase        string `json:"phase"` // "BUY_PHASE" | "COMBAT" | "POST_ROUND"
}

type PlayerRealtimeStats struct {
	PUUID                 string  `json:"puuid"`
	RiotID                string  `json:"riotId"`
	AgentName             string  `json:"agentName"`
	AgentIconURL          string  `json:"agentIconUrl"`
	Kills                 int     `json:"kills"`
	Deaths                int     `json:"deaths"`
	Assists               int     `json:"assists"`
	KDRatio               float64 `json:"kdRatio"`
	CombatScore           int     `json:"combatScore"`
	EconomyCredits        int     `json:"economyCredits"`
	CurrentTierName       string  `json:"currentTierName"`
	CurrentTierIconURL    string  `json:"currentTierIconUrl"`
	RankingRating         int     `json:"rankingRating"`
	RRChangeLastMatch     int     `json:"rrChangeLastMatch"`
	MapWinRate            float64 `json:"mapWinRate"`
	HistoricalMatchesOnMap int    `json:"historicalMatchesOnMap"`
}

type OverlayTelemetryPayload struct {
	Timestamp   int64               `json:"timestamp"`
	MatchState  ValorantMatchState  `json:"matchState"`
	PlayerStats PlayerRealtimeStats `json:"playerStats"`
}

// --- Comprehensive Historical Analytics (Web Dashboard & Deep View) ---

type AgentMastery struct {
	AgentName      string  `json:"agentName"`
	AgentIconURL   string  `json:"agentIconUrl"`
	MatchesPlayed  int     `json:"matchesPlayed"`
	WinRate        float64 `json:"winRate"`
	AvgCombatScore float64 `json:"avgCombatScore"`
	KDRatio        float64 `json:"kdRatio"`
}

type WeaponMarksmanship struct {
	WeaponName     string  `json:"weaponName"`
	TotalKills     int     `json:"totalKills"`
	HeadshotPercent float64 `json:"headshotPercent"`
	BodyshotPercent float64 `json:"bodyshotPercent"`
	LegshotPercent  float64 `json:"legshotPercent"`
}

type MapPerformance struct {
	MapName       string  `json:"mapName"`
	MatchesPlayed int     `json:"matchesPlayed"`
	WinRate       float64 `json:"winRate"`
	AttackWinRate float64 `json:"attackWinRate"`
	DefendWinRate float64 `json:"defendWinRate"`
}

type PlayerHistoricalSummary struct {
	PUUID            string               `json:"puuid"`
	RiotID           string               `json:"riotId"`
	CurrentRank      string               `json:"currentRank"`
	PeakRank         string               `json:"peakRank"`
	OverallKDRatio   float64              `json:"overallKdRatio"`
	OverallWinRate   float64              `json:"overallWinRate"`
	TotalMatches     int                  `json:"totalMatches"`
	AgentMasteries   []AgentMastery       `json:"agentMasteries"`
	WeaponAccuracy   []WeaponMarksmanship `json:"weaponAccuracy"`
	MapMatrix        []MapPerformance     `json:"mapMatrix"`
	LastUpdated      time.Time            `json:"lastUpdated"`
}
